### PR TITLE
Rendering pipeline

### LINKED ISSUE #3: build the starting structure of the rendering pipeline
We will need a basic structure setup along with all of the boiler plate code implemented for the rendering pipeline with vulkan.

### LINKED ISSUE #4: implement a windowing class that can connect to the rendering pipeline

### PR DESCRIPTION
**Pull request by Rinzii**

### CODE REVIEW

**Comment by Rinzii** [comment]
I'd like a review of how I'm handling instance creation inside of the device class. relevant commit is a13d01e72d861eb3b5d63cd5b4d0472a183d0366

**Comment by Rinzii** [comment]
Also light review of the vkHelpers is also fine

File: editor/src/main.cpp [hunk 1]
```diff
@@ -5,7 +5,10 @@
 #include <logger/log.hpp>
 
 int main() {
+	// We don't want to initialize the logger in release mode.
```
Comment (karnkaul) [inline editor/src/main.cpp]: Why not?

File: engine/include/application.hpp [hunk 2]
```diff
@@ -11,14 +14,24 @@ class Application {
...
+
+	// Initialization functions
+
+	void initializeGraphics();
+
 public:
-	static constexpr u32 m_width = 800;
-	static constexpr u32 m_height = 600;
+	static const u32 m_width = 800;
+	static const u32 m_height = 600;
```
Comment (karnkaul) [inline engine/include/application.hpp]: Why not `constexpr`?

File: engine/include/application.hpp [hunk 3]
```diff
@@ -11,14 +14,24 @@ class Application {
...
 
 private:
-	Window m_window{ 800, 600, "Genesis Engine" };
+	Window m_window { m_width, m_height, "Genesis Engine" };
+	GraphicsDevice m_graphicsDevice { m_window };
+	GraphicsPipeline m_graphicsPipeline { m_graphicsDevice , GraphicsPipeline::defaultPipelineConfigInfo(m_width, m_height)};
```
Comment (karnkaul) [inline engine/include/application.hpp]: Why is there a pipeline already?

File: engine/include/graphics/apiVersion.hpp
Comment (karnkaul) [inline engine/include/graphics/apiVersion.hpp:1]: Why is this in graphics?

File: engine/include/graphics/device.hpp [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+    {
+        vk::SurfaceCapabilitiesKHR capabilities;
+        std::vector<vk::SurfaceFormatKHR> formats;
+        std::vector<vk::PresentModeKHR> presentModes;
+    };
+
+	struct QueueFamilyIndices
+    {
+        std::optional<u32> graphicsFamily;
+        std::optional<u32> presentFamily;
```
Comment (karnkaul) [inline engine/include/graphics/device.hpp]: Are you sure you want to get into multi-queue right now? It's 10x more complicated than dealing with a single queue.

File: engine/include/graphics/device.hpp [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+		{
+			return querySwapChainSupport(m_physicalDevice);
+		}
+
+		u32 findMemoryType(u32 typeFilter, vk::MemoryPropertyFlags properties);
+
+		QueueFamilyIndices findPhysicalQueueFamilies()
+		{
+			return findQueueFamilies(m_physicalDevice);
+		};
```
Comment (karnkaul) [inline engine/include/graphics/device.hpp]: Bunch of commented code being checked in

File: engine/include/graphics/device.hpp [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+							std::vector<std::string> const & layers,
+							std::vector<std::string> const & extensions,
+							u32 const & apiVersion
+		);
+
+		//void createDebugMessenger();
+		//void createSurface();
+		//void pickPhysicalDevice();
+		//void createLogicalDevice();
+		//void createCommandPool();
```
Comment (karnkaul) [inline engine/include/graphics/device.hpp]: More commented code

File: engine/include/graphics/device.hpp [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+		//bool checkDeviceExtensionSupport(vk::PhysicalDevice device);
+		//SwapChainSupportDetails querySwapChainSupport(vk::PhysicalDevice device);
+
+
+
+		vk::Instance m_instance;
+		vk::DebugUtilsMessengerEXT m_debugMessenger;
+		vk::PhysicalDevice m_physicalDevice;
+		vk::Device m_device;
+		vk::SurfaceKHR m_surface;
```
Comment (karnkaul) [inline engine/include/graphics/device.hpp]: Why aren't we using `vk::UniqueInstance` etc?

File: engine/include/graphics/device.hpp [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+
+		vk::PhysicalDeviceProperties m_physicalDeviceProperties;
+
+		vk::Queue m_graphicsQueue;
+		vk::Queue m_presentQueue;
+		vk::CommandPool m_commandPool;
```
Comment (karnkaul) [inline engine/include/graphics/device.hpp]: What's this for?

File: engine/include/graphics/device.hpp [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+
+		Window &m_window; // NOLINT(cppcoreguidelines-avoid-const-or-ref-data-members)
```
Comment (karnkaul) [inline engine/include/graphics/device.hpp]: Why does the graphics device need a persistent reference to the window?

File: engine/include/graphics/device.hpp [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+
+		const std::vector<std::string> validationLayers = {"VK_LAYER_KHRONOS_validation"};
+		const std::vector<std::string> deviceExtensions = {VK_KHR_SWAPCHAIN_EXTENSION_NAME};
```
Comment (karnkaul) [inline engine/include/graphics/device.hpp]: Why are these members (instead of being `static` or even internal to the cpp file)?

File: engine/include/graphics/pipeline.hpp [hunk 11]
```diff
@@ -1,20 +1,55 @@
...
 
-    class Pipeline {
+#include "device.hpp"
+
+namespace gen
+{
+	struct PipelineConfigInfo
+	{
+
+	};
```
Comment (karnkaul) [inline engine/include/graphics/pipeline.hpp]: What's this?

File: engine/include/graphics/pipeline.hpp [hunk 12]
```diff
@@ -1,20 +1,55 @@
...
+
+    class GraphicsPipeline {
     public:
-        Pipeline(const std::string& vertFilePath, const std::string& fragFilePath);
+		GraphicsPipeline(GraphicsDevice &device, const PipelineConfigInfo& configInfo);
+		~GraphicsPipeline();
+
+		GraphicsPipeline(const GraphicsPipeline&) = delete;
+		GraphicsPipeline& operator=(const GraphicsPipeline&) = delete;
+
```
Comment (karnkaul) [inline engine/include/graphics/pipeline.hpp]: We don't need pipelines for a while.

File: engine/include/graphics/pipeline.hpp [hunk 13]
```diff
@@ -1,20 +1,55 @@
...
+		// NOLINTNEXTLINE The assumption with this is that the device will outlive the pipeline at all times since a pipeline fundamentally needs a pipeline to exist
+		GraphicsDevice& m_device;
+		vk::Pipeline m_graphicsPipeline;
+		vk::ShaderModule m_vertShaderModule;
+		vk::ShaderModule m_fragShaderModule;
+
+		const PipelineConfigInfo m_configInfo;
+
+		vk::Instance m_instance;
+		vk::DebugUtilsMessengerEXT m_debugMessenger;
```
Comment (karnkaul) [inline engine/include/graphics/pipeline.hpp]: Unrelated code in PR

File: engine/include/util/version.hpp
Comment (karnkaul) [inline engine/include/util/version.hpp:1]: Why are there two Version types?

File: engine/include/windowing/window.hpp [hunk 14]
```diff
@@ -8,41 +8,52 @@
 
 // TODO: Replace this with a proper implementation
 
-// Forward declare GLFWwindow
-struct GLFWwindow;
+// Start Forward declarations
+
+#include <GLFW/glfw3.h>
+
+#include <vulkan/vulkan.hpp>
```
Comment (karnkaul) [inline engine/include/windowing/window.hpp]: This is incorrect, `GLFW/glfw3.h` needs to be included _after_ `GLFW_INCLUDE_VULKAN` has been defined.

File: engine/include/windowing/window.hpp [hunk 15]
```diff
@@ -8,41 +8,52 @@
...
+	vk::SurfaceKHR createWindowSurface(vk::Instance & instance);
 
 
-	// TODO: Currently these functions will break the window.
+	// Getters
+
+	GEN_NODISCARD inline int getWidth() const { return m_width; }
+	GEN_NODISCARD inline int getHeight() const { return m_height; }
+	GEN_NODISCARD inline const std::string& getTitle() const { return m_title; }
+	GEN_NODISCARD inline GLFWwindow* getHandle() const { return m_window; }
```
Comment (karnkaul) [inline engine/include/windowing/window.hpp]: Member functions are already implicitly `inline`.

File: engine/include/windowing/window.hpp [hunk 16]
```diff
@@ -8,41 +8,52 @@
...
 	GEN_NODISCARD bool shouldClose() const;
 	static void pollEvents() ;
 
+	// Helper functions
 
-public:
-	GEN_NODISCARD inline int GetWidth() const { return m_width; }
-	GEN_NODISCARD inline int GetHeight() const { return m_height; }
-	GEN_NODISCARD inline const std::string& GetTitle() const { return m_title; }
+	vk::SurfaceKHR createWindowSurface(vk::Instance & instance);
```
Comment (karnkaul) [inline engine/include/windowing/window.hpp]: Not using `vk::UniqueSurfaceKHR` here

File: engine/include/windowing/window.hpp [hunk 17]
```diff
@@ -8,41 +8,52 @@
...
-	// TODO: Currently these functions will break the window.
+	// Getters
+
+	GEN_NODISCARD inline int getWidth() const { return m_width; }
+	GEN_NODISCARD inline int getHeight() const { return m_height; }
+	GEN_NODISCARD inline const std::string& getTitle() const { return m_title; }
+	GEN_NODISCARD inline GLFWwindow* getHandle() const { return m_window; }
+
+
+	// TODO: Currently these functions will break the window. Have not yet implemented window resizing.
```
Comment (karnkaul) [inline engine/include/windowing/window.hpp]: I'd remove functions that break the app. `main` should be stable and without dangling stuff hanging around.

File: engine/src/application.cpp [hunk 18]
```diff
@@ -7,9 +7,24 @@
...
+            Window::pollEvents();
+        }
+
+	}
+	void Application::shutdown() {
+	}
+
+	void Application::initializeGraphics() {
+
+	}
```
Comment (karnkaul) [inline engine/src/application.cpp]: What's the point of these empty functions?

File: engine/src/graphics/device.cpp [hunk 19]
```diff
@@ -0,0 +1,355 @@
...
+	{
+		// TODO: Allow for the vulkan api to be specified in a config file or through command line arguments.
+		// We could also dynamically check for the highest supported version of the vulkan api, but that feels outside the scope of this project.
+		createInstance(window.getTitle(), "Genesis Engine", validationLayers, {}, VK_API_VERSION_1_0);
+
+	    //createDebugMessenger();
+        //createSurface();
+		//pickPhysicalDevice();
+		//createLogicalDevice();
+		//createCommandPool();
```
Comment (karnkaul) [inline engine/src/graphics/device.cpp]: More commented out code

File: engine/src/graphics/device.cpp [hunk 20]
```diff
@@ -0,0 +1,355 @@
...
+#include <GLFW/glfw3.h>
+
+namespace gen
+{
+
+	GraphicsDevice::GraphicsDevice(gen::Window & window) : m_window{window}
+	{
+		// TODO: Allow for the vulkan api to be specified in a config file or through command line arguments.
+		// We could also dynamically check for the highest supported version of the vulkan api, but that feels outside the scope of this project.
+		createInstance(window.getTitle(), "Genesis Engine", validationLayers, {}, VK_API_VERSION_1_0);
```
Comment (karnkaul) [inline engine/src/graphics/device.cpp]: Why are we going with 1.0?? We should require 1.3.

File: engine/src/graphics/device.cpp [hunk 21]
```diff
@@ -0,0 +1,355 @@
...
+	GraphicsDevice::~GraphicsDevice()
+	{
+		m_device.destroyCommandPool(m_commandPool);
+        m_device.destroy();
+
+		//NOLINTNEXTLINE validation layers are enabled conditionally and this trips up clang-tidy
+		if (enableValidationLayers) { m_instance.destroyDebugUtilsMessengerEXT(m_debugMessenger); }
+
+		m_instance.destroySurfaceKHR(m_surface);
+		m_instance.destroy();
```
Comment (karnkaul) [inline engine/src/graphics/device.cpp]: Unecessary with `vk::Unique*`.

File: engine/src/graphics/device.cpp [hunk 22]
```diff
@@ -0,0 +1,355 @@
...
+		createInstance(window.getTitle(), "Genesis Engine", validationLayers, {}, VK_API_VERSION_1_0);
+
+	    //createDebugMessenger();
+        //createSurface();
+		//pickPhysicalDevice();
+		//createLogicalDevice();
+		//createCommandPool();
+	}
+
+	GraphicsDevice::~GraphicsDevice()
```
Comment (karnkaul) [inline engine/src/graphics/device.cpp:25]: Missing `device.waitIdle()` before destruction of stuff

File: engine/src/graphics/device.cpp [hunk 23]
```diff
@@ -0,0 +1,355 @@
...
+		auto vkGetInstanceProcAddr = vkdl.getProcAddress<PFN_vkGetInstanceProcAddr>("vkGetInstanceProcAddr");
+		VULKAN_HPP_DEFAULT_DISPATCHER.init(vkGetInstanceProcAddr);
+#endif
+
+		if (enableValidationLayers) //NOLINT validation layers are enabled conditionally and this trips up clang-tidy
+        {
+			if (!vk::util::checkValidationLayerSupport(validationLayers))
+			{
+				gen::logger::error("vulkan", "validation layers requested, but not available!");
+				throw std::runtime_error("validation layers requested, but not available!");
```
Comment (karnkaul) [inline engine/src/graphics/device.cpp]: I don't think this is a fatal error, the app can run fine without validation layers, just disable them.

File: engine/src/graphics/device.cpp [hunk 24]
```diff
@@ -0,0 +1,355 @@
...
+#endif
+		);
+
+		m_instance = vk::createInstance(vk::util::makeInstanceCreateInfoChain(appInfo, enabledLayers, enabledExtensions).get<vk::InstanceCreateInfo>());
+
+#if (VULKAN_HPP_DISPATCH_LOADER_DYNAMIC == 1)
+		// initialize function pointers for instance
+		VULKAN_HPP_DEFAULT_DISPATCHER.init(m_instance);
+#endif
+	}
```
Comment (karnkaul) [inline engine/src/graphics/device.cpp]: Such macro-infested code is largely unreadable, I have no idea what's going on...

File: engine/src/graphics/device.cpp [hunk 25]
```diff
@@ -0,0 +1,355 @@
...
+        {
+            if (((typeFilter & (1 << i)) != 0U) && (memProperties.memoryTypes[i].propertyFlags & properties) == properties)
+            {
+                return i;
+            }
+        }
+
+		gen::logger::error("vulkan", "failed to find suitable memory type!");
+        throw std::runtime_error("failed to find suitable memory type!");
+	}
```
Comment (karnkaul) [inline engine/src/graphics/device.cpp]: 300 lines of commented code

File: engine/src/graphics/pipeline.cpp [hunk 26]
```diff
@@ -1,29 +1,72 @@
...
+        createGraphicsPipeline();
     }
 
-    std::vector<char> Pipeline::readFile(const std::string & filePath) {
+	GraphicsPipeline::~GraphicsPipeline()
+	{
+		destroyGraphicsPipeline();
+	}
+
+    std::vector<char> GraphicsPipeline::readFile(const std::string & filePath)
```
Comment (karnkaul) [inline engine/src/graphics/pipeline.cpp:25]: Why is this function in `GraphicsPipeline`?

File: engine/include/graphics/vkHelpers.hpp [hunk 27]
```diff
@@ -0,0 +1,55 @@
...
+#endif
+	makeInstanceCreateInfoChain( vk::ApplicationInfo const & appInfo,
+                                 std::vector< const char * > const & layers,
+                                 std::vector< const char * > const & extensions );
+
+	vk::DebugUtilsMessengerCreateInfoEXT makeDebugUtilsMessengerCreateInfoEXT();
+
+	bool checkValidationLayerSupport(const std::vector<std::string>& validationLayers);
+
+	bool checkDeviceExtensionSupport(vk::PhysicalDevice device, const std::vector<const char*>& deviceExtensions);
```
Comment (karnkaul) [inline engine/include/graphics/vkHelpers.hpp]: Why are these members, and if they need access to data members, why aren't they `const`?

File: engine/src/graphics/pipeline.cpp
Comment (karnkaul) [inline engine/src/graphics/pipeline.cpp:1]: Out of scope changes in PR

File: engine/src/graphics/vkHelpers.cpp [hunk 28]
```diff
@@ -0,0 +1,242 @@
...
+		return true;
+#endif
+	}
+
+	// Check if the required device extensions are supported by the given physical device.
+	bool checkDeviceExtensionSupport(vk::PhysicalDevice device, const std::vector<const char*>& deviceExtensions)
+	{
+		auto availableExtensions = device.enumerateDeviceExtensionProperties();
+
+		std::set<std::string> requiredExtensions(deviceExtensions.begin(), deviceExtensions.end());
```
Comment (karnkaul) [inline engine/src/graphics/vkHelpers.cpp]: `unordered_set` will be more efficient.

File: engine/src/graphics/vkHelpers.cpp [hunk 29]
```diff
@@ -0,0 +1,242 @@
...
+		}
+
+		return true;
+#else
+		return true;
+#endif
+	}
+
+	// Check if the required device extensions are supported by the given physical device.
+	bool checkDeviceExtensionSupport(vk::PhysicalDevice device, const std::vector<const char*>& deviceExtensions)
```
Comment (karnkaul) [inline engine/src/graphics/vkHelpers.cpp]: Could take `std::span` instead

File: engine/src/graphics/vkHelpers.cpp [hunk 30]
```diff
@@ -0,0 +1,242 @@
...
+		compiler.set_entry_point("main", executionModel);
+
+		std::basic_string<char, std::char_traits<char>, std::allocator<char>> spirvData = compiler.compile();
+		spvShader = std::vector<unsigned int>(spirvData.begin(), spirvData.end());
+
+		return true;
+	}
+
+	// Create a Vulkan shader module from HLSL shader code.
+	vk::ShaderModule createShaderModule( vk::Device const & device, vk::ShaderStageFlagBits shaderStage, std::string const & shaderText )
```
Comment (karnkaul) [inline engine/src/graphics/vkHelpers.cpp]: Should return `vk::UniqueShaderModule`

File: engine/src/logger/log.cpp [hunk 31]
```diff
@@ -1,20 +1,33 @@
...
+    }
 
-void logger::debug(std::string_view category, std::string_view message, SrcLoc const& location) {
-	Instance::print(message, Context::make(category, Level::eDebug, location));
-}
+    void logger::debug(std::string_view category, std::string_view message, SrcLoc const& location)
+	{
+        Instance::print(message, Context::make(category, Level::eDebug, location));
+    }
+#else // If GEN_NDEBUG is defined then diable logging. The creation of this symbol is created by cmake.
```
Comment (karnkaul) [inline engine/src/logger/log.cpp]: Why don't we want any logs in Release??

File: engine/src/windowing/window.cpp [hunk 32]
```diff
@@ -2,29 +2,34 @@
...
 #include <GLFW/glfw3.h>
 
+
 #include <utility>
 
 namespace gen {
 
     Window::Window(int w, int h, std::string title) : m_width{w}, m_height{h}, m_title{std::move(title)} { // NOLINT(cppcoreguidelines-pro-type-member-init)
-        initWindow();
+        init();
```
Comment (karnkaul) [inline engine/src/windowing/window.cpp]: Why is this a separate function if it's only ever called once in the constructor?

File: engine/src/windowing/window.cpp [hunk 33]
```diff
@@ -33,6 +38,16 @@ namespace gen {
 
     }
 
-
+	vk::SurfaceKHR Window::createWindowSurface(vk::Instance & instance)
```
Comment (karnkaul) [inline engine/src/windowing/window.cpp]: Should return `vk::UniqueSurfaceKHR`.

**Review by karnkaul — CHANGES_REQUESTED** [review]

**Comment by karnkaul** [comment]
Why are all checks consistently failing?

**Comment by karnkaul** [comment]
The title of the PR is misleading, this work sets up a Vulkan instance, no pipelines or rendering.

File: engine/include/windowing/window.hpp (hunk 16 above)
Comment (karnkaul) [reply engine/include/windowing/window.hpp]: Also not using nodiscard

File: engine/src/logger/log.cpp (hunk 31 above)
Comment (Rinzii) [reply engine/src/logger/log.cpp]: My thought process is you likely don’t want to be outputting every log message possible include just simple debug info and information on that is being logged. Though after thinking about it I think a better approach would be to instead log only errors and maybe warnings in release as that could be pertinent to the user.

File: engine/include/graphics/device.hpp (hunk 4 above)
Comment (Rinzii) [reply engine/include/graphics/device.hpp]: All the resources I’ve been using has been using multi but if its far more complicated then I’ll prob just pull it back to a single queue.

File: engine/include/graphics/apiVersion.hpp
Comment (Rinzii) [reply engine/include/graphics/apiVersion.hpp:1]: This was something I was experimenting to allow to switching between graphics API’s but have kinda abandoned at the moment while I focused on getting the renderer actually properly working. Even though I want to allow for configuration of the rendering api I don’t like the way I was trying so I’ll likely remove it.

File: editor/src/main.cpp (hunk 1 above)
Comment (Rinzii) [reply editor/src/main.cpp]: Reasoning mentioned in other convo about logging.

**Comment by Rinzii** [comment]
> The title of the PR is misleading, this work sets up a Vulkan instance, no pipelines or rendering.

This PR is a draft that was originally meant to be reviewed for the entire rendering pipeline but since I’ve brought the work load to have more reviews it is no longer correct. I just reused it instead of closing the PR and opening a new one though I likely should of just closed and restarted it.

**Comment by Rinzii** [comment]
> Why are all checks consistently failing?

No PCH issues. My bad on my part forgot to test without PCH. Will correct all of those issues before re-review!

File: engine/include/graphics/device.hpp (hunk 5 above)
Comment (Rinzii) [reply engine/include/graphics/device.hpp]: Commented because I did not want to remove it completely. Will be valid once I move onto the next step.

File: engine/include/graphics/device.hpp (hunk 6 above)
Comment (Rinzii) [reply engine/include/graphics/device.hpp]: Commented because I did not want to remove it completely. Will be valid once I move onto the next step.

File: engine/include/graphics/device.hpp (hunk 4 above)
Comment (karnkaul) [reply engine/include/graphics/device.hpp]: Which resources use multiple queues? [vkguide](https://vkguide.dev/docs/chapter-1/vulkan_commands_code/#grabbing-the-queue) uses only one, so does Vulkan tutorial.

> We are requesting a graphics queue, which supports all we need for the guide.

File: engine/include/graphics/pipeline.hpp (hunk 11 above)
Comment (Rinzii) [reply engine/include/graphics/pipeline.hpp]: This was added for for future use allowing for pipeline configuration.
